src/data/right_inv.lean

```diff
@@ -0,0 +1,64 @@
+/-
+Copyright © 2021 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import tactic.basic
+
+/-!
+# Bundled right inverse (section)
+
+In this file we implement the bundled version of a right inverse function. The final purpose of
+this type is to conveniently implement sections of bundles.
+
+## Implementation notes
+
+Let `E : B → Type*` be the collection of fibers of a bundle on `B`.
+In the case of sections of bundles we prove that the type of right inverses of `bundle.proj E` is
+equivalent to the type `Π x : B, E x`. Both implementations of sections have their advantages:
+`right_inv (proj E)` is better to talk about properties such as continuity, whereas `Π x : B, E x`
+works better with algebraic structures: it naturally inherits the algebraic structures present on
+the fibers, thanks to the `Π` instances. It is hence a good idea to implement both and write down
+the equivalence between them, but then we have to choose only one for our final implementation of
+sections. We chose `right_inv (proj E)` because it is more general (some of the results in the
+theory of sections are true for `right_inv f`, for a generic `f`), and it has a more natural
+coercion to a function `B → bundle.total_space E` which is finally what we want. Note that we could
+also register a coercion to `B → bundle.total_space E` from `Π x : B, E x`, and it is not a bad idea
+to implement things this way, but it is just slightly more clumsy: in the case of continuous
+sections, in the first case one can implement continuous sections as
+`structure continuous_section (f : α → β) extends right_inv f, C(β, α)`
+whereas in the second way one would have to write continuity manually; it is not a big deal but the
+first way is slightly more natural. Also note that to recover the value of the section on the fiber
+in the first way one can use `.2` whereas the second way would make use of coercions.
+
+-/
+
+/-- Bundled right inverse of a function. Note that here the `nolint` has a true mathematical
+meaning: the structure is inhabited iff the function is surjective. -/
+@[nolint has_inhabited_instance]
+structure right_inv {α: Type*} {β: Type*} (f : α → β) :=
+(to_fun : β → α)
+(right_inv' : function.right_inverse to_fun f)
+
+namespace right_inv
+
+variables {α: Type*} {β: Type*} {f : α → β} {g h : right_inv f}
+
+instance : has_coe_to_fun (right_inv f) (λ _, β → α) := ⟨right_inv.to_fun⟩
+
+@[simp] lemma to_fun_eq_coe (g : right_inv f) : g.to_fun = ⇑g := rfl
+
+lemma coe_injective (H : (g : β → α) = h) : g = h := /- Why is ⇑g not working? -/
+by { cases g, cases h, congr' }
```

Comment:
Weird behavior since `out_param` was introduced in `has_coe_to_fun`. Should I ignore it?

Reply:
You should provide a `fun_like` instance instead, which includes this lemma as a field.

Reply:
I tried but thigs do not work if I do that. For example the following code
```
instance right_inv_fun_like : fun_like (right_inv f) β (λ _, α) :=
⟨right_inv.to_fun, λ g h H, by { cases g, cases h, congr' }⟩

@[simp] lemma to_fun_eq_coe (g : right_inv f) : g.to_fun = ⇑g := rfl
```
does not work because `⇑g` does not get recognized properly.

Reply:
Generally you should not have to write `⇑` yourself anyway - just make sure that lean inserts it in the place you expect when your cursor is on the `rfl`.

You should add the `fun_like` instance anyway, even though it didn't change the `⇑` weirdness you were seeing.

Reply:
I can try but I need help because it breaks my code in other points. If I do that, when I write `ext` it tells me
```
no applicable extensionality rule found for right_inv
```
Can you guess already what the problem is (I guess this could have appeared in many places when refactoring the library)? Or do you want an mwe?

Reply:
@eric-wieser do you have an idea of what might be going on here?

Reply:
Have you read through the docstring near `fun_like` which explain what lemmas you're required to write alongside it?

Reply:
https://leanprover-community.github.io/mathlib_docs/data/fun_like/basic.html#basic-usage-of-codefun_likecode

Reply:
No actually I did not notice there was all that documentation, now I will fix it! But at this point I wonder what is the advantage of `fun_like` if one has to write `ext` and `has_coe_to_fun` anyways...

Reply:
You get `congr_fun` and `congr_arg` for free, for one.